---
title: "deseq2"
output: html_document

DESeq2 is a very useful and versatile package for assessing differential gene expression in a dataset. Please feel free to check out the [documentation](https://bioconductor.org/packages/release/bioc/manuals/DESeq2/man/DESeq2.pdf) as well as this quick-start [guide](https://bioc.ism.ac.jp/packages/2.14/bioc/vignettes/DESeq2/inst/doc/beginner.pdf) to find out more about the full power of this package.


Let's begin by loading in DESeq2, as well as some other important packages that we will be using in our analysis. 

```{r}
# run this in the console to install the packages you need:

#if (!requireNamespace("BiocManager", quietly = TRUE))
#install.packages("BiocManager")

#BiocManager::install("DESeq2")

#install.packages("gplots")


```


```{r}

library("DESeq2")

library("ggplot2")

library("RColorBrewer")

```

Let's now import our counts matrix. Since we already generated a "clean counts" matrix, let's just import that one:

```{r }

counts <- read.csv('Mouse_LPS_counts_for_deseq2.csv',
                  header=TRUE, row.names=1)
head(counts)

```

Let's also import the conditions matrix. We will need both for out differential expression analysis:

```{r}

col_data <- read.csv('Mouse_LPS_conditions_for_deseq2.csv',
                  header=TRUE, row.names=1)

head(col_data)

```

Let's now make a DESeq Dataset as an object. This will take into account out counts and condition data, as well as experimental design (experimental variable being set as condition):

```{r}
dds <- DESeqDataSetFromMatrix(countData = counts,
                              colData = col_data,
                              design = ~ condition)
```

Let's run DESeq analysis and summarize the results with res:

```{r}
dds <- DESeq(dds)
```

```{r}

res <- results(dds)

```

```{r}

res

```
```{r}

summary(res)

```

Let's make a graphical representation of our results given our cutoffs. In an MA plot, the x-axis is the average expression over all samples for a particular gene, while the y-axis the log2 fold change between treatment and control:

```{r}

plotMA(res, main="DESeq2", ylim=c(-2,2))

```

We also have the option of adjusting our parameters of significance. Say, we wished to look at events with an padj value < 0.05. We could do the following:

```{r}
res05 <- results(dds, alpha=0.05)

plotMA(res05, main="alpha=0.05", ylim=c(-2,2))

```

Another useful quality metric is principal component analysis (PCA). This is a useful way of measuring sample-to-sample variance between and within experimental groups. This method transforms high dimensional data into a set of linearly uncorrelated variables that can then be represented on a 2D (or higher dimension) plane. Each principal component defines a degree of variance, the largest of which is represented in principal component 1 (PC1), then PC2, etc. For a more detailed intro, please check out the following [notebook](https://github.com/ryanmarina/BMS_bioinformatics_bootcamp_2018/blob/master/tutorials/Intro_to_PCA.ipynb) that Josh prepared.

```{r}

rld <- rlog(dds)

vsd <- varianceStabilizingTransformation(dds)

data <- plotPCA(rld, intgroup="condition", returnData=TRUE)
percentVar <- round(100 * attr(data, "percentVar"))
ggplot(data, aes(PC1, PC2, color=condition)) +
    geom_point(size=3) +
    xlab(paste0("PC1: ",percentVar[1],"% variance")) 
    ylab(paste0("PC2: ",percentVar[2],"% variance"))

```

Let's take our transformed data and

```{r}

sampleDists <- dist(t(assay(rld)))
sampleDistMatrix <- as.matrix(sampleDists)
rownames(sampleDistMatrix) <- paste(rld$condition)
colnames(sampleDistMatrix) <- paste(rld$condition)
colors <- colorRampPalette( rev(brewer.pal(9, "Blues")) )(255)
heatmap(sampleDistMatrix,
clustering_distance_rows=sampleDists,
clustering_distance_cols=sampleDists,
col=colors)

```

We can also cluster variable genes in the same way that we clustered samples. To do so, we will take only the most variable genes in our dataset (let's say the top 200 variably expressed genes). We can then make a clustered heatmap based upon the fold change of our normalized read counts:

```{r}

library( "gplots" )
library( "genefilter" )
topVarGenes <- head( order( rowVars( assay(rld) ), decreasing=TRUE ), 200 )

```

```{r}

#par(mar=c(.1,.1,.1,.1))
#png(filename='test.png', width=1000, height=1000)
heatmap.2( assay(rld)[ topVarGenes, ], scale="row", 
     trace="none", dendrogram="row", margins=c(6,1),
     col = colorRampPalette( rev(brewer.pal(9, "RdBu")) )(255), cexRow=1,cexCol=1, srtCol = 45)

```

Lastly, let's export our DESeq2 data table. From this, we can determine which genes are differentially expressed and what they mean with respect to our biological processes.

```{r}

write.csv(as.data.frame(res), file="dendritic_differential_expression.csv")

```

